rejex.mjs

```diff
@@ -0,0 +1,172 @@
+// This file implements a complete regular expression engine, as given by the
+// formal definition of Regular Expressions at
+// https://en.wikipedia.org/wiki/Regular_expression#Formal_definition
+//
+// In particular, it implements:
+//  - concatenation (using the class ConcatExpr)
+//  - alternation (using the class AlternationExpr)
+//  - Kleene star (using the class RepetitionExpr)
+//
+// Using these three operations, all other kinds of regular expressions can be
+// expressed.
+//
+// For simplicity, no parser is provided for the regex *language*. Instead, a
+// regular expression is constructed as a tree of objects consisting of:
+//  - CharMatchExpr
+//  - ConcatExpr
+//  - AlternationExpr
+//  - RepetitionExpr
+//
+// When given a tree of such objects, the regexMatch function below can tell
+// you if a given string does, or does not match the given regular expression
+// object tree.
+//
+// A very simple example of this would be:
+//
+//   regexMatch(str, new CharMatchExpr("h"))
+//
+// which matches strings which contain exclusively the single character "h".
+//
+// See the unit tests in rejex.test.js for more examples.
+
+
+// Match a single character
+export class CharMatchExpr {
+  constructor(chr) {
+    this.chr = chr
+  }
+
+  // Match against a list of character tokens.
```

Comment:
> aren't all these match function comments the same

They are. If I had an abstract base class I'd put it there, but see my other comment about avoiding too many new OOP concepts. So I opted to duplicate the documentation instead of using an ABC.

> why call them "character tokens"?

This was intentional, and I want to talk about it during the talk. Regular Expressions aren't limited to matching against characters. It can be used against any stream of data, bytes, tokens, whatever. Tokens here is a placeholder used commonly in parsers to refer to a single "thing" without ascribing its datatype.

Consider, for example, a regex that runs against the *bytes* of a file, which could be used to write regexes against binary data. In that case, a token is a byte.

Or another example, a regex over the token output of a [lexer](https://en.wikipedia.org/wiki/Lexical_analysis), which would treat constructs like `if` as a single token.

Reply:
That's a great point! Can mention how it might be used in the script interpreter/compilation?